fenicsadapter/fenicsadapter.py

```diff
@@ -1,759 +1,331 @@
-        if self._dimension == 2:
-            for i in range(self._vals.ndim):
-                return_value[i] = self._f[i](x[0], x[1])
-        if self._dimension == 3:
-            for i in range(self._vals.ndim):
-                return_value[i] = self._f[i](x[0], x[1], x[2])
-        return return_value
-
-
-class ExactInterpolationExpression(CustomExpression):
-    """Uses cubic spline interpolation for implementation of CustomExpression.interpolate. Only allows interpolation on
-    coupling that are parallel to the y axis, and if the coordinates in self._coords_y are ordered such that the nodes
-    on the coupling mesh are traversed w.r.t their connectivity.
-    However, this method allows to exactly recover the solution at the coupling interface, if it is a polynomial of
-    order 3 or lower.
-    See also https://github.com/precice/fenics-adapter/milestone/1
-    """
-    def create_interpolant(self):
-        interpolant = []
-        if self._dimension == 2:
-            if self.is_scalar_valued():  # check if scalar or vector-valued
-                interpolant.append(interp1d(self._coords_y, self._vals, bounds_error=False, fill_value="extrapolate", kind="cubic"))
-            elif self.is_vector_valued():
-                interpolant.append(interp1d(self._coords_y, self._vals[:, 0].flatten(), bounds_error=False, fill_value="extrapolate", kind="cubic"))
-                interpolant.append(interp1d(self._coords_y, self._vals[:, 1].flatten(), bounds_error=False, fill_value="extrapolate", kind="cubic"))
-            else:
-                raise Exception("Problem dimension and data dimension not matching.")
-        else:
-            raise Exception("Dimension of the function is invalid/not supported.")
+    def update_point_sources(self, data):
+        """
+        Update values of point sources using new data
+        This function only works for 2D-pseudo3D coupling.
+        :param data: 2D data from preCICE
+        :return:
+        """
+        return get_forces_as_point_sources(self._Dirichlet_Boundary, self._function_space, self._coupling_mesh_vertices,
+                                           data)
 
-        return interpolant
+    def read(self):
+        """
+        Reads data from preCICE. Depending on the dimensions of the simulation (2D-3D Coupling, 2D-2D coupling or
+        Scalar/Vector write function) read_data is converted.
+        Note: For quasi 2D fenics in a 3D coupled simulation the y component of the vectors is deleted.
+        :return: data read from preCICE in the form of a numpy array with the values like it is used by preCICE
+        """
+        assert (self._read_function_type in list(FunctionType))
 
-    def interpolate(self, x):
-        assert ((self.is_scalar_valued() and self._vals.ndim == 1) or
-                (self.is_vector_valued() and self._vals.ndim == self._dimension))
+        read_data = convert_fenics_to_precice(self._read_function, self._coupling_mesh_vertices)
```

Comment:
Here, `self._read_function` is used, but `read_data` is not used a directly being overwritten a few lines below.